cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -121,6 +122,84 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
       static_cast<const CountOptions&>(*args.options));
 }
 
+// ----------------------------------------------------------------------
+// Distinct Count implementation
+
+template <typename Type>
+struct CountDistinctImpl : public ScalarAggregator {
+  using MemoTable = typename arrow::internal::HashTraits<Type>::MemoTableType;
+
+  explicit CountDistinctImpl(MemoryPool* memory_pool, CountOptions options)
+      : options(std::move(options)), memo_table_(new MemoTable(memory_pool, 0)) {}
+
+  Status Consume(KernelContext*, const ExecBatch& batch) override {
+    if (batch[0].is_array()) {
+      const ArrayData& arr = *batch[0].array();
+      auto visit_null = [&]() {
+        if (this->nulls > 0) return Status::OK();
+        ++this->nulls;
+        return Status::OK();
+      };
+      auto visit_value = [&](typename Type::c_type arg) {
+        int y;
+        RETURN_NOT_OK(memo_table_->GetOrInsert(arg, &y));
+        return Status::OK();
+      };
+      RETURN_NOT_OK(VisitArrayDataInline<Type>(arr, visit_value, visit_null));
+      this->non_nulls += this->memo_table_->size();
+    } else {
+      const Scalar& input = *batch[0].scalar();
+      this->nulls += !input.is_valid * batch.length;
+      this->non_nulls += input.is_valid * batch.length;
+    }
```

Comment:
@edponce Can you tell me how to write a test for this case? I would like to fix this by using a C++ test, so I can understand better the error.
btw thanks good catch!

Reply:
It should not be possible for `batch.length != 1` for `Scalars` because [the length of the batch is set to `1` when constructing the `ExecBatch`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/exec.h#L202-L211), so no change nor test needed for this.

Reply:
I think that comment is misleading/incorrect based on how the query engine/scanner have developed. You could have an ExecBatch formed by projecting just the partition columns from a batch, in which case you'd have scalar rows with a length > 1.

Reply:
In fact, all this was explicitly added to support that case. Can you file a JIRA to update that comment?

Reply:
*all this being the `batch.length` usage in the aggregate kernels, which previously mishandled this case.

Reply:
Created ARROW-14203 to fix comment

Reply:
Thanks!